### ISSUE TITLE
[FEA] CUDA 12 pip wheels should throw a helpful warning or error in cases where JIT MVC will not work

### ISSUE THREAD

**Issue by vyasr** [rapidsai/cudf#13339]
**Is your feature request related to a problem? Please describe.**
cudf currently packages ptx files to support string UDFs. The ptx packaging is necessary because we need to define a translation layer between libcudf C++ string APIs and numba's Python machinery, which only understands C APIs (no C++ name mangling allowed) and is based on compiling Python code down to ptx. However, ptx does not support minor version compatibility, so a cudf built using a particular CTK will not run correctly if the driver is older. 

Moreover, since UDFs require JIT-compilation at runtime, cudf UDFs will also fail to run correctly if the CUDA runtime present at runtime is older than the CUDA driver at runtime. This affects all UDFs, not just string UDFs.

For CUDA 11, we work around these issues using the ptxcompiler and cubinlinker packages. The solution for CUDA 12 will be to use the new nvjitlink library, but we will not be supporting this for our initial release of CUDA 12 wheels in 23.06. What we should do in 23.06 is ensure that users do not see opaque errors when they are in an unsupported environment.

**Describe the solution you'd like**
We should add logic in cudf that checks whether the driver is older than the runtime when the runtime > 12.0. Since we're compiling with CUDA 12.0.1 in our wheel builds we don't need to test anything else there. I would suggest that we include this error check in the [_compile_or_get](https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/udf/utils.py#L286) function.

**Additional context**
This is the short term patch while we wait for the longer-term solution with nvjitlink, which will resolve https://github.com/rapidsai/cudf/issues/12822.

**Comment by brandon-b-miller** [comment]
Just to confirm we should be o.k if driver==runtime==12.0 right? 

I think I can handle this as part of https://github.com/rapidsai/cudf/pull/13337.

**Comment by vyasr** [comment]
Yup, and I'd be fine with you tossing this into #13337 (that's almost 1337 :smile:). Yes, I think that is fine. I believe that this basically boils down to adding a check to `compile_udf` that does `if cuda.runtime.get_version() > (12, 0) and cuda.runtime.get_version() > cuda.driver.get_version(): raise ...`.

**Comment by vyasr** [comment]
CC @gmarkall, let us know if that sounds right to you.

**Comment by gmarkall** [comment]
Yes, that sounds right.
